### DISCUSSION TITLE
Faster convolutions

### DISCUSSION THREAD

**Discussion by OverLordGoldDragon** [kymatio/kymatio#753] [General]
Padding to x16 of input length is computationally intractable. It's also wasteful: the widest (in time) filter dictates pad length for all filters, where many can live with x2. 

### 1. Variable length wavelets

I propose varying the length of each wavelet, taking only as much as needed: `len(psi) == len(x) + supp(psi)/2` (but such that permits max subsampling). Then we do `ifft(xpf * psi_f)`, where `xpf = fft(x_pad[slice])`:

  - Cost of extra `fft` is less than savings from shorter dot-multiply and `ifft`.
  - Number of FFTs on slices can be reduced by grouping wavelets by length rather than varying every single one.

### 2. ~FFT~ convolution

For narrowest wavelets, having support x20-40 less than unpadded input's, FFT convolution is suboptimal. Supports below a threshold can be sampled exactly at support's length, and convolved linearly (can be [strided](https://stackoverflow.com/a/48098534/10133797)) - also without per-length overhead (extra `fft`).

### 3. Periodic-reflect padding & upsample-trimmed wavelets

It's possible to get `reflect` to form a perfect loop, akin to `periodic` padding, which is equivalent to infinite padding in circular convolution. The problem here is wavelet distortion:

![image](https://user-images.githubusercontent.com/16495490/121823970-c6b99800-ccb9-11eb-9563-427e295d51bd.png)

Left is wavelet sampled at convolution length, right is sampled at longer then trimmed, as described [here](https://github.com/OverLordGoldDragon/ssqueezepy/issues/29). Right is proper. While this resolves shape distortion, it doesn't address failure to decay to 0, which is the main distortion. 

Not much can be done here to make output identical with full padding. A heuristic is an "energy correction term", or perhaps re-normalizing right's wavelet - that may bring error to acceptable margin. 

In the worst case of full padding, `periodic-reflect` slashes required padding in half [like zero](https://github.com/kymatio/kymatio/pull/734).

### 4. Reflecting intermediate outputs

Necessary for 1 and 2 since we still require full length for subsequent compute (`psi2`, `phi`). Assuming full padding that avoids all boundary effects, convolution outside signal bounds with `reflect` padding will also be reflected (after modulus). 

Example pipeline is thus, going per 1:

 1. `x_pad = pad_reflect_to_max_length(x)`
 2. `xpf = fft(x_pad[slice])`, unless already done for this length
 3. `c = subsample_fourier(xpf * psi)`
 4. `r = ifft(c)`
 5. `U = modulus(r)`
 6. `U = pad_reflect(U)`
 7. `c = fft(U)`
 8. `c = subsample_fourier(c * phi)`
 9. `r = ifft(c)`
 10. `S = unpad(r)`

so that's two extra computations, 2 and 6, that should still net-save relative to full length conv. Fortunately for the `T == N` case we can skip 6 per `mean()`. For small `T` we re-pad less (or none with `average=False`), more prominently for `F` & `average_fr=False`.

<hr>

### Plan

I'll be trying 1 & 4, keeping it to dyadic lengths for simplicity. 2 & 3 are rather complicated and better left for standalone overhauls.

**Comment by OverLordGoldDragon** [comment]
Green = halfway marker; purple = `N`, i.e. minimum padding minus 1

![image](https://user-images.githubusercontent.com/16495490/123526782-29b02380-d6eb-11eb-9ab5-8ab04a40bb8b.png)

For high and low `Q` and `J`, more than half of all filters' width is below input's length. Convoltions can thus be split in two, one padding to `log2_N + 1` and other to max - but it's only worth the bother for large `J`.

**Reply by OverLordGoldDragon** [reply]
Problem with repadding:

![image](https://user-images.githubusercontent.com/16495490/123551819-64b26580-d784-11eb-984f-c75cfe7ac7d0.png)

Blue: signal -- red: pad boundaries -- purple, green: kernel at different shifts -- red dots: strided conv -- 2nd plot: conv output

Reflected strided != strided reflect. However, all necessary information is contained within padding bounds: one approach thus is to pad with `ifft(fft_upsample(strided_conv))[offset::stride]`, where `offset` above would be `1` for right padding. I assume `offset` can be done in freq domain to avoid `ifft` on upsampled. But I wonder if there's a faster way? @lostanlen

**Reply by lostanlen** [reply]
can you explain the problem from the start?

**Reply by OverLordGoldDragon** [reply]
Actually I think the proposed scheme will be fast enough; `fft_upsample` doesn't do `fft`. Question for now is, how do we do `psi_t[1::2]` in freq domain? I notice its `abs(fft)` is same as `psi_t[0::2]`'s but I can't tell how the phase changes.

**Reply by OverLordGoldDragon** [reply]
So far I've looked to periodization expressions; seems a solution can be derived by solving for `t = ns + s` instead

![image](https://user-images.githubusercontent.com/16495490/123554929-97178f00-d793-11eb-9d96-f4880b823d7e.png)

**Reply by OverLordGoldDragon** [reply]
Figured it out:

```python
g = np.random.randn(128)
g[31:-31] = 0  # unaliased case

g0 = fft(ifft(g)[0::2])
g1 = fft(ifft(g)[1::2])
# find g1 (g2) from g0
w = np.arange(len(g0)//2 + 1) / len(g)
w = np.hstack([w, -w[1:-1][::-1]])
g2 = g0 * np.exp(1j * 2*np.pi * w)

assert np.allclose(g1, g2)
```
It manifests as a phase correction term when replacing `ns` with `(n + 1)*s` in the Poisson formula. I'll test whether this generalizes for any `offset` and subsampling factor.

**Reply by OverLordGoldDragon** [reply]
It fails for aliased. Phase correction must be done _before_ folding:

```python
def subsample(x, k, offset=0):
    assert offset < 2**k
    w = np.arange(len(x)//2 + 1) / len(x)
    w = np.hstack([w, -w[1:-1][::-1]])
    e = np.exp(1j * 2*np.pi * w * offset)

    xfs = (e*fft(x)).reshape(2**k, -1).mean(axis=0)
    return ifft(xfs)

x = np.random.randn(128)
for k in (1, 2, 3):
    for offset in range(2**k):
        assert np.allclose(x[offset::2**k], subsample(x, k, offset))
```
This means we can't do `offset` correction at subsampled length and must dot-multiply at unsubsampled (but can do unpadded). Also `abs(fft)` for different `offset`s are _not_ equal in aliased case.

### Complexity comparison

 1. `U_c = psi_f * xf16` -- `xf16 == fft([x]*16)` (`x` padded to 16x length) -- dot-multiply at x16
 2. `U_h = U_c.reshape(2, -1).mean(axis=0)`  (`subsample_fourier(k=1)`) -- mean at x8
 3. `U_c = ifft(U_h)` -- `ifft` at x8
 4. `U = modulus(U_c)` -- `abs(x)` at x8

vs proposed:

 1. `U_c = psi_f * xf2` -- dot-multiply at x2
 2. `U_h = U_c.reshape(2, -1).mean(axis=0)` -- mean at x2
 3. `U_c = ifft(U_h)` -- `ifft` at x1
 4. `U_m = modulus(U_c)` -- `abs(x)` at x1
 5. `U_c0 = psi_f0 * xf1` -- dot-multiply at x1 (unpadded `x`)
 6. `U_h0 = (e*U_c0).reshape(2, -1).mean(axis=0)` -- dot-multiply at x1, mean at x0.5
 7. Possibly repeat 6 for different `offset` at left / right -- or reuse `U_h1 = U_h0`
 8. `U_c0 = ifft(U_h0)` -- `ifft` at x0.5
 9. `U_m0 = modulus(U_c0)` -- modulus at x0.5
 10. Possibly repeat 8 & 9
 11. `U_u = unpad(U_m)` -- view at x1 --> nothing
 12. `U_re = concat([U_m0, U_m, U_m1])` -- allocating x8 array with repeated reflections of `U_c0` and `U_c1`

Note 5 can be replaced with `fft_upsample` which is just padding in Fourier (asm. no alias), effectively "nothing".

Every other op's insignificant relative to FFT so I'll just compare FFTs:

 - **Current**: 8
 - **Proposed**: 1 + 0.5*2 = 2

At surface, `*2` isn't the worst case: every reflection might require its own `offset`. There are `2**k` possible `offset`s, but we already have `1` from original `U_m`. Padding to x8 requires 4 reflections on each side (rather 3.5 but still 4 offsets), so a total of `min(2**k - 1, 8 - 1)` IFFTs. _However_: we at most need one IFFT that's unsubsampled, this IFFT at x1, which we can then slice with any `offset` and `k`, which makes the worst case again a `1` (=`0.5*2`). Thus:

 - **Proposed**: 1.5 to 2

Since there's a bunch of other stuff going on I'll just call it 2.25 - thus a gain of **x3.6**, assuming FFT complexity is linear. Yet my benches on _dyadic_ `N` show sub-linear dependence, which actually makes the gain less. Regardless it's significant.

I'll be benching these to be sure.

**Reply by OverLordGoldDragon** [reply]
Seems to work: an x1.5-1.6 overall speedup for `max_order=1` time scattering with max `J`. Large kernel convs continue to dominate compute time; the only improvements I can see there is point 3 in original post. Without losing too much, an easy speedup is to lower max padding by 1: x16 -> x8, that brings total speedup to x3 relative to all x16 convs, but is still x3.8 slower than all x2 convs.

I'm hoping to find another improvement, but maybe x1.6 is worth it...

**Reply by OverLordGoldDragon** [reply]
Actually the speedup is x2 with `average=False`; it was still averaging with max kernel lowpass, `average_global` isn't implemented for Scattering1D. So with `True` and lowpass also padded optimally the speedup might be around x1.8. Further, the large kernel convs actually comprise only 25% of compute time... guess subsampling's carrying hard there.

With `average_fr_global` or `average_fr=False` we won't even need to repad in freq scattering... it'll be _much_ faster.

**Reply by OverLordGoldDragon** [reply]
x3.1 speedup without stride offset correction, a x1.5 gain... I'm not sure how important this correction is, for one the 'incorrect' variant is actually more reflect-ive. The difference in coeffs is also quite small - perhaps a kwarg to control this.

**Reply by OverLordGoldDragon** [reply]
So... unpadding creates DC and negative frequencies:

![image](https://user-images.githubusercontent.com/16495490/124049796-4f5e6500-da2a-11eb-8109-0a9d3e8d0952.png)

I've not yet looked into what happens after phi but I expect the effect to be likewise small. This small effect regardless manifests as aliasing and prohibits upsampling for repadding:

![image](https://user-images.githubusercontent.com/16495490/124050325-66ea1d80-da2b-11eb-8e44-467e3e3f70ba.png)

Unpadding before upsampling is thus not viable, must do _after_ ... which is more expensive: it's IFFT at unsubsampled padded length. This defeats the purpose of subsampling entirely - but not of repadding.

It's also counterintuitive how subsampling a full signal can be lossless while subsampling a _portion_ of the same signal be lossy... but makes sense: the "portion" contains less information, and signal outside the portion could be anything, so recovery is non-unique. (put differently the loss is from unpadding)

Looking at unpadding as subsampling in Fourier (for approx unaliased case), we can predict loss: in above case it's easy to see the "spectrum's spectrum" will have components in Nyquist octave beyond just phase info.

**Comment by OverLordGoldDragon** [comment]
Shorter wavelets and repadding yield an x2 speedup across majority of configurations, and can be pushed a bit further - but benefits drop sharply by simply reducing max padding from x16 to x8, which is tolerable. There's many possibilities, maybe it can be close to x2 with x8 - but it's too much work.

Commit state containing faster convs that passed `allclose` for Scattering1D, and in-progress for JTFS, [here](https://github.com/lostanlen/kymatio/pull/15/commits/7cdb16536dbe5829686ebc61efd587a6f339fc33).

I cease my efforts on this.

**Reply by lostanlen** [reply]
@OverLordGoldDragon thank you so much for looking into this! it's a good point of reference to have
